File: ios/apps/AutoTester/AutoTester/testCases/CartoDBLightTestCase.m
//
//  CartoDBLightTestCase.m
//  AutoTester
//
//  Created by Steve Gifford on 3/15/16.
//  Copyright © 2016-2017 mousebird consulting. All rights reserved.
//

#import "WhirlyGlobeComponent.h"
#import "CartoDBLightTestCase.h"

@implementation CartoDBLightTestCase

- (instancetype)init
{
    if (self = [super init]) {
        self.name = @"CartoDB Light";
        self.captureDelay = 5;
		self.implementations = MaplyTestCaseImplementationMap | MaplyTestCaseImplementationGlobe;
    }
    
    return self;
}

- (void) setupBaseLayer:(MaplyBaseViewController *) baseLayer
{
    NSString * baseCacheDir = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) objectAtIndex:0];
    NSString * cartodbTilesCacheDir = [NSString stringWithFormat:@"%@/cartodbtiles/", baseCacheDir];
    int maxZoom = 22;
    MaplyRemoteTileSource *tileSource = [[MaplyRemoteTileSource alloc] initWithBaseURL:@"http://light_all.basemaps.cartocdn.com/light_all/" ext:@"png" minZoom:0 maxZoom:maxZoom];
    tileSource.cacheDir = cartodbTilesCacheDir;
    MaplyQuadImageTilesLayer *layer = [[MaplyQuadImageTilesLayer alloc] initWithCoordSystem:tileSource.coordSys tileSource:tileSource];
    if ([baseLayer isKindOfClass:[WhirlyGlobeViewController class]]) {
        layer.handleEdges = true;
        layer.coverPoles = true;
    }
    else {
        layer.handleEdges = false;
        layer.coverPoles = false;
    }
    layer.requireElev = false;
    layer.waitLoad = false;
    layer.singleLevelLoading = false;
    layer.drawPriority = 0;
    [baseLayer addLayer:layer];
}

- (void)setUpWithMap:(MaplyViewController *)mapVC
{
    [self setupBaseLayer:mapVC];
    mapVC.height = 0.7;
    [mapVC animateToPosition:MaplyCoordinateMakeWithDegrees(8.4666749, 47.3774337) time:1.0];
}

- (void)setUpWithGlobe:(WhirlyGlobeViewController *)globeVC
{
    [self setupBaseLayer:globeVC];
    [globeVC animateToPosition:MaplyCoordinateMakeWithDegrees(8.4666749, 47.3774337) time:1.0];
}

@end

File: ios/apps/AutoTester/AutoTester/testCases/CartoDBLightTestCase.swift
//
//  GlobeSamplerTestCase.swift
//  AutoTester
//
//  Created by Stephen Gifford on 3/27/18.
//  Copyright © 2018 mousebird consulting.
//

import UIKit

class CartoDBLightTestCase: MaplyTestCase {
    
    override init() {
        super.init()
        
        self.name = "CartoDB Light Test Case"
        self.implementations = [.globe,.map]
    }
    
    var imageLoader : MaplyQuadImageLoader? = nil
    
    // Put together a quad sampler layer
    func setupLoader(_ baseVC: MaplyBaseViewController) -> MaplyQuadImageLoader? {
        let cacheDir = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
        let thisCacheDir = "\(cacheDir)/cartodblight/"
        let maxZoom = Int32(16)
        let tileInfo = MaplyRemoteTileInfoNew(baseURL: "http://basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}@2x.png",
                                              minZoom: Int32(0),
                                              maxZoom: Int32(maxZoom))
        tileInfo.cacheDir = thisCacheDir
        
        // Parameters describing how we want a globe broken down
        let sampleParams = MaplySamplingParams()
        sampleParams.coordSys = MaplySphericalMercator(webStandard: ())
        sampleParams.coverPoles = true
        sampleParams.edgeMatching = true
        sampleParams.minZoom = tileInfo.minZoom()
        sampleParams.maxZoom = tileInfo.maxZoom()
        sampleParams.singleLevel = true
        
        guard let imageLoader = MaplyQuadImageLoader(params: sampleParams, tileInfo: tileInfo, viewC: baseVC) else {
            return nil
        }
        imageLoader.imageFormat = .imageUShort565;
        //        imageLoader.debugMode = true
        
        return imageLoader
    }
    
    override func setUpWithGlobe(_ globeVC: WhirlyGlobeViewController) {
        imageLoader = setupLoader(globeVC)
    }
    
    override func setUpWithMap(_ mapVC: MaplyViewController) {
        imageLoader = setupLoader(mapVC)
    }
}
